File: 05397.py
class Node:
    def __init__(self, c, prev, next):
        self.c = c
        self.prev = prev
        self.next = next


def main():
    t = int(input())
    for _ in range(t):
        string = input()
        head = Node('', None, None)
        current = head
        for c in string:
            if c not in '-<>':
                current.next = Node(c, current, current.next)
                current = current.next
                if current.next:
                    current.next.prev = current
            elif c == '<' and current.c:
                current = current.prev
            elif c == '>' and current.next:
                current = current.next
            elif c == '-' and current.c:
                current = current.prev
                current.next = current.next.next
                if current.next:
                    current.next.prev = current
        output = ''
        current = head
        while current:
            output += current.c
            current = current.next
        print(output)


main()

File: 05397.cpp
#include <cstdio>
#include <cstdlib>

struct node{
	char c;
	node *prev;
	node *next;
};

using namespace std;

int main(){
	int t;
	scanf("%d", &t);
	for(int a = 0; a < t; ++a){
		char string[1000001];
		scanf("%s", string);
		node *head = (node *)malloc(sizeof(node));
		head->c = ' ';
		node *current = head;
		for(int i = 0; string[i] != '\0'; ++i){
			char c = string[i];
			if(c != '<' && c != '>' && c != '-'){
				node *temp = (node *)malloc(sizeof(node));
				temp->c = c;
				temp->prev = current;
				temp->next = current->next;
				current->next = temp;
				current = current->next;
				if(current->next != NULL){
					current->next->prev = current;
				}
			}else if(c == '<' && current->c != ' '){
				current = current->prev;
			}else if(c == '>' && current->next != NULL){
				current = current->next;
			}else if(c == '-' && current->c != ' '){
				node *temp = current;
				current = current->prev;
				current->next = current->next->next;
				if(current->next != NULL){
					current->next->prev = current;
				}
				free(temp);
			}
			current = head->next;
			while(current != NULL){
				printf("%c", current->c);
				node *temp = current;
				current = current->next;
				free(temp);
			}
			printf("\n");
		}
	}
	return 0;
}
